File: backend/src/main/java/br/com/calciolari/datahub/imports/domain/parser/ParserInput.java
package br.com.calciolari.datahub.imports.domain.parser;

import java.io.InputStream;
import java.util.Objects;
import java.util.Optional;

/**
 * Bytes + optional non-authoritative metadata for {@link ImportParser}.
 * {@code originalFilename} is never identity.
 */
public record ParserInput(
		InputStream content,
		long contentLength,
		String originalFilename,
		String detectedType
) {
	public ParserInput {
		Objects.requireNonNull(content, "content");
		if (contentLength < 0) {
			throw new IllegalArgumentException("contentLength must be >= 0");
		}
	}

	public Optional<String> originalFilenameOptional() {
		return Optional.ofNullable(originalFilename).filter(name -> !name.isBlank());
	}
}

File: backend/src/Calciolari.DataHub/Imports/Domain/Parser/ParserInput.cs
namespace Calciolari.DataHub.Imports.Domain.Parser;

/// <summary>
/// Bytes + optional non-authoritative metadata for <see cref="IImportParser"/>.
/// originalFilename is never identity.
/// </summary>
public sealed record ParserInput(
    Stream Content,
    long ContentLength,
    string? OriginalFilename,
    string? DetectedType)
{
    public string? OriginalFilenameOptional() =>
        string.IsNullOrWhiteSpace(OriginalFilename) ? null : OriginalFilename;

    public static ParserInput Create(Stream content, long contentLength, string? originalFilename, string? detectedType)
    {
        ArgumentNullException.ThrowIfNull(content);
        if (contentLength < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(contentLength), "contentLength must be >= 0");
        }

        return new ParserInput(content, contentLength, originalFilename, detectedType);
    }
}
